pyci/rdm/conditions.py

```diff
@@ -0,0 +1,66 @@
+# This file is part of PyCI.
+#
+# PyCI is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your
+# option) any later version.
+#
+# PyCI is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+# for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with PyCI. If not, see <http://www.gnu.org/licenses/>.
+
+r"""PyCI RDM Conditions module."""
+
+import numpy as np
+
+__all__ = [
+    "find_closest_sdp",
+]
+
+def find_closest_sdp(dm, constraint, alpha):
+    r"""
+    Projection onto a semidefinite constraint.
+
+    Parameters
+    ----------
+    dm : np.ndarray
+        Density matrix.
+    constraint : function
+        Positive semidefinite constraint, linear mapping.
+    alpha : float
+        Value of the correct trace.
+
+    """
+    #symmetrize if necessary
+    L = constraint(dm) + constraint(dm).conj().T
+    #find eigendecomposition
+    vals, vecs = np.linalg.eig(L)
+    #calculate the shift, sigma0
+    sigma0 = calculate_shift(vals, alpha)
+    
+    #calculate the closest semidefinite positive matrix with correct trace
+    L_closest = vals @ np.diag(vecs - sigma0) @ vecs.conj().T
```

Comment:
I think this is correct, considering that numpy.linalg.eigh returns the eigenvectors as columns. Just to be sure, can you reconstruct the (random symmetric) matrix from it's eigenvalues and eigenvectors using this formula?

Reply:
Yes we can, as far as I checked

Reply:
@RichRick1 has a good test. That's something we can automate to be sure that things work correctly. I.e., with $sigma_0 = 0$ the original matrix should be returned. With $sigma_0$ optimized the matrix should be positive semidefinite with the right trace. Testing this for a few random input matrices (with the right trace) is a good way to ensure the code is working correctly.